src/Internal/Payment/PaymentContext.php

```diff
@@ -244,14 +262,71 @@ public function get_intent(): ?WC_Payments_API_Abstract_Intention {
 		return $this->get( 'intent' );
 	}
 
+	/**
+	 * Updates state transition and associated changes
+	 *
+	 * @param string $state The state.
+	 */
+	public function update_state_transition( string $state ): void {
+		$this->transitions[ $state ] = $this->changes;
+		$this->changes               = [];
+	}
+
+	/**
+	 * Returns the transitions and changes as a string that can be logged.
+	 *
+	 * @return string
+	 */
+	public function log_changes(): string {
+		$log            = '';
+		$previous_state = null;
+		foreach ( $this->transitions as $state => $changes ) {
+			$log           .= $previous_state ?
+								"Transition from '" . $previous_state . "' to '" . $state :
+								"Payment initialized in '" . $state;
+			$log           .= "' {" . PHP_EOL;
+			$log           .= implode( PHP_EOL, $this->changes_to_str( $changes ) ) . PHP_EOL;
+			$log           .= '}' . PHP_EOL;
+			$previous_state = $state;
+		}
+		if ( ! empty( $this->changes ) ) {
+			$log .= "Changes within '" . $previous_state . "' {" . PHP_EOL;
+			$log .= '}' . PHP_EOL;
+		}
+		return $log;
+	}
+
+	/**
+	 * Returns the changes array as a string that can be logged.
+	 *
+	 * @param array $changes Array of Change objects.
+	 *
+	 * @return array
+	 */
+	private function changes_to_str( $changes ) : array {
+		$changes_string = array_map(
+			function( Change $change ) {
+				if ( $change->get_old_value() ) {
+					$str = '  Changed ' . $change->get_key() . ' from ' . wp_json_encode( $change->get_old_value() ) . ' to ' . wp_json_encode( $change->get_new_value() );
```

Comment:
Let's add JSON_PRETTY_PRINT here, and expand this.

Note: That might requrie adding some tabs before new lines, to make it consistent.

Reply:
Added JSON_PRETTY_PRINT and adjusted the log formatting

Reply:
We still need to indent correctly. Right now the log looks like this, and it's hard to understand where things start and end: 
<img width="1120" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/5311119/5868ab4b-71b3-4c8e-b67e-2bc77d5c4c12">